### PR TITLE
BWAN-790

### PR DESCRIPTION
**Pull request by praveenfff**
FRR code to create and process the egress updates
1. Added hook in the BGP code to process the egress update received from the controller
2. created new message between BGP and Zebra to handle the egress updates
3. Handled the egress update in the zebra and send to data- plane.

### CODE REVIEW

File: bgpd/bgp_zebra.c [hunk 1]
```diff
@@ -2705,6 +2706,35 @@ static int bgp_pbr_get_ifnumber(struct bgp *bgp)
 	return cnt;
 }
 
+struct egress_data {
```
Comment (rohitfff) [inline bgpd/bgp_zebra.c]: struct egress_data is redefined  , recommend define it once in a header file

File: zebra/zapi_msg.c [hunk 2]
```diff
@@ -2342,6 +2345,32 @@ static inline void zread_ipset_entry(ZAPI_HANDLER_ARGS)
...
+{
+	uint32_t size = 0;
+	uint32_t dest= 0;
+	struct stream *s;
+	s = msg;
+	STREAM_GETL(s, size);
+	STREAM_GETL(s, dest);
+	struct infiot_egress_hook* update = NULL;
+	update = (struct infiot_egress_hook*)malloc(sizeof(struct infiot_egress_hook));
+	update->dest = dest;
```
Comment (rohitfff) [inline zebra/zapi_msg.c:2363]: missing null check after malloc.

File: zebra/zebra_fpm.c [hunk 3]
```diff
@@ -1221,12 +1377,37 @@ static inline int zfpm_is_enabled(void)
 	return zfpm_g->enabled;
 }
 
+int zfpm_trigger_egress_update(struct infiot_egress_hook *upd, const char *reason)
+{
+	if (!zfpm_conn_is_up()) {
+		return 0;
+	}
+	infiot_egress *info = malloc(sizeof(infiot_egress));
+	for(int i=0;i<upd->size;i++){
```
Comment (rohitfff) [inline zebra/zebra_fpm.c:1390]: missing malloc null check

**Review by rohitfff — APPROVED** [review]

File: bgpd/bgp_zebra.c [hunk 4]
```diff
@@ -2705,6 +2706,35 @@ static int bgp_pbr_get_ifnumber(struct bgp *bgp)
...
+	int ret = 0;
+	s = zclient->obuf;
+	stream_reset(s);
+	zclient_create_header(s,
+			      TRUE ? ZEBRA_INFIOT_EGRESS_ADD :
+			      ZEBRA_INFIOT_EGRESS_DELETE,
+			      VRF_DEFAULT);
+	stream_putl(s, size);
+	stream_putl(s, info.destination);
+	for(int i=0;i<size;i++) {
```
Comment (rohitfff) [inline bgpd/bgp_zebra.c]: Minor request, coding convention, missing spaces Example. VSS Code or other IDEs fix this automatically 

for(int i=0;i<size;i++)      <- Current
for(int i=0; i<size; i++)    <- With Space

File: bgpd/bgp_zebra.c (hunk 1 above)
Comment (praveenfff) [reply bgpd/bgp_zebra.c]: Ack

File: bgpd/bgp_zebra.c (hunk 4 above)
Comment (praveenfff) [reply bgpd/bgp_zebra.c]: Ack

File: zebra/zapi_msg.c (hunk 2 above)
Comment (praveenfff) [reply zebra/zapi_msg.c:2363]: Added

File: zebra/zebra_fpm.c (hunk 3 above)
Comment (praveenfff) [reply zebra/zebra_fpm.c:1390]: Added

**Pull request merged**